### PR TITLE
change `addCSourceFiles` to use `LazyPath` instead `Dependency`

### PR DESCRIPTION
**Pull request by Rexicon226**
cc @jacobly0 

Makes it possible to prepend a LazyPath that might have been derived from a `zig-cache` dep, and prepend it to a list of files.

## Use Case:

I was adding `libgit2` as a dep to my project by downloading the tar ball from Github archive and since, (submodules bad) I used the build system. This leaves me with a `LazyPath` to the location in `zig-cache`, and a large list of files I need to add to the exe with the same flags. Since `addCSourceFiles` takes a "hardcoded" list of relative paths, I had no way to access things inside the `zig-cache`

### CODE REVIEW

File: lib/std/Build/Step/Compile.zig [hunk 1]
```diff
@@ -1197,15 +1197,11 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
...
 
-                        if (c_source_files.dependency) |dep| {
-                            for (c_source_files.files) |file| {
-                                try zig_args.append(dep.builder.pathFromRoot(file));
-                            }
-                        } else {
-                            for (c_source_files.files) |file| {
-                                try zig_args.append(b.pathFromRoot(file));
-                            }
+                        const root_path = c_source_files.root.getPath2(b, step);
```
Comment (jacobly0) [inline lib/std/Build/Step/Compile.zig:1200]:
```suggestion
                        const root_path = c_source_files.root.getPath2(b, module.owner);
```

No idea what the difference is, but that's what ever other call to `getPath2` in this function does.

**Comment by jacobly0** [comment]
# Migration Guide

```
exe.root_module.addCSourceFiles(.{
    .dependency = my_dep,
    .files = &.{
        "first.c",
        "second.c",
     },
    .flags = &.{
        "-DMY_DEF",
    },
});
```
&downarrow;
```
exe.root_module.addCSourceFiles(.{
    .root = my_dep.path(""),
    .files = &.{
        "first.c",
        "second.c",
     },
    .flags = &.{
        "-DMY_DEF",
    },
});
```

Calls to `addCSourceFiles` that don't use `dependency` are not affected.

File: lib/std/Build/Module.zig [hunk 2]
```diff
@@ -79,7 +79,7 @@ pub const SystemLib = struct {
 };
 
 pub const CSourceFiles = struct {
-    dependency: ?*std.Build.Dependency,
+    root: LazyPath = .{ .path = "" },
```
Comment (jacobly0) [inline lib/std/Build/Module.zig:82]:
```suggestion
    root: LazyPath,
```
This default belongs in the other struct, forgetting to pass that value to this struct should be an error.

File: lib/std/Build/Module.zig [hunk 3]
```diff
@@ -79,7 +79,7 @@ pub const SystemLib = struct {
...
     /// If `dependency` is not null relative to it,
```
Comment (jacobly0) [inline lib/std/Build/Module.zig:83]: This comment needs updating.

**Review by jacobly0 — CHANGES_REQUESTED** [review]

**Comment by andrewrk** [comment]
Why does it need to be a breaking change?

**Comment by Rexicon226** [comment]
> Why does it need to be a breaking change?

Changes name from `dependency` to `root`.

**Comment by castholm** [comment]
This might need to address the same safety concerns as `LazyPath.dirname`. Does this implementation protect against navigating outside of root directories?

```zig
exe.root_module.addCSourceFiles(.{
    .root = my_dep.path(""),
    .files = &.{
        "../../../../../../../secret.c",
     },
});
```

It doesn't block this PR, but I also think it would be useful if this logic was extended to some general `LazyPath.join` or `LazyPath.relative` function, i.e. `fn join(self: LazyPath, other: []const u8) LazyPath`. I have encountered several use cases where I've wanted to do the inverse of `LazyPath.dirname` and navigate further down into a generated directory, for example when running system commands that create a directory structure with files (with no control of where individual files go) and needing to pick out a specific file from that directory.

**Comment by jacobly0** [comment]
> Why does it need to be a breaking change?

I would say that it doesn't technically need to be breaking, other than the fact that having having both `root` and `artifact` in the struct allows calling the function with invalid arguments, whereas removing `artifact` prevents an invalid set of arguments in the type system.  Making it a tagged union doesn't help because that would be a breaking change anyway.

On the other hand, you could also prevent this from being a breaking change by just creating a new function and having the old function trivially call the new function.

**Review by jacobly0 — APPROVED** [review]

**Comment by kcbanner** [comment]
I'm seeing an issue on windows where C source paths are now being resolved incorrectly like this: `C:\Users\kcbanner\projects\ion\c:\Users\kcbanner\projects\ion\deps/zmesh/libs/meshoptimizer/simplifier.cpp`. 

I bisected it to `3eacd1b2e56728a291b4e5dc443a56fa0b4cab14`. I don't have a reduction yet, but the relevant build.zig.zon snippet (where deps/zmesh is https://github.com/zig-gamedev/zig-gamedev/tree/main/libs/zmesh):

```zig
    .dependencies = .{
...
        .zmesh = .{ .path = "deps/zmesh" },
````

**Comment by jacobly0** [comment]
`zmesh` passes absolute paths to `addCSourceFiles` which is not supported.

With a few changes
```
const zmesh_package = @import("zmesh").package(b, target, optimize, .{ .args = .{
    .shape_use_32bit_indices = shape_use_32bit_indices,
    .shared = shared,
} });
_ = zmesh_package.zmesh;
_ = zmesh_package.zmesh_options;
_ = zmesh_package.zmesh_c_cpp;
```
should be able to be completely replaced with
```
const zmesh_package = b.dependency("zmesh", .{
    .target = target,
    .optimize = optimize,
    .shape_use_32bit_indices = shape_use_32bit_indices,
    .shared = shared,
});
_ = zmesh_package.module("zmesh");
_ = zmesh_package.module("zmesh_options");
_ = zmesh_package.artifact("zmesh_c_cpp");
```

**Comment by nektro** [comment]
> `zmesh` passes absolute paths to `addCSourceFiles` which is not supported.

this should be asserted

**Comment by kcbanner** [comment]
> ```
> const zmesh_package = b.dependency("zmesh", .{
>     .target = target,
>     .optimize = optimize,
>     .shape_use_32bit_indices = shape_use_32bit_indices,
>     .shared = shared,
> });
> exe.root_module.addImport("zmesh", zmesh_package.module("zmesh"));
> exe.root_module.addImport("zmesh_options", zmesh_package.module("zmesh_options"));
> exe.linkLibrary(zmesh_package.artifact("zmesh_c_cpp"));
> const zmesh_tests = &b.addRunArtifact(zmesh_package.artifact("zmesh-tests")).step;
> ```

For this technique to work, I had to make some changes:

- Had to install `zmesh-tests` and `zmesh` (aka zmesh_c_cpp). This is the only way to get `Dependency.artifact` to see these artifacts. This is a bit inconvenient as the tests exe and static library end up in zig-out, when it would be ideal to run / link these directly from where they are in the cache. Is there a way to get at a dependency's CompileSteps for direct linking?
- Manually give zmesh_options a name (this seems fine, since manually exposing them isn't that common)
```
    const zmesh_options = step.createModule();
    _ = b.addModule("zmesh_options", .{
        .root_source_file = step.getOutput(),
    });
```

**Comment by nektro** [comment]
> * Is there a way to get at a dependency's CompileSteps for direct linking?

is that not what `exe.linkLibrary(zmesh_package.artifact("zmesh_c_cpp"));` does?

**Comment by kcbanner** [comment]
> > Is there a way to get at a dependency's CompileSteps for direct linking?
> 
> is that not what `exe.linkLibrary(zmesh_package.artifact("zmesh_c_cpp"));` does?

Only if you've `b.getInstallStep().dependOn(&b.addInstallArtifact(zmesh_c_cpp` in the dependencies build script. This results in the library being installed into zig-out/lib, which is a different behavior then before.

**Comment by nektro** [comment]
imo that should filed as a new bug

**Pull request merged**